### ISSUE TITLE
Spatial Transformer Networks

### ISSUE THREAD

**Issue by saraansh1999** [mlpack/mlpack#2250]
This is a great feature to be included for vision tasks.
Resource: [https://arxiv.org/pdf/1506.02025.pdf](url)

Implementing this would also require a grid sampler.

Is any of this implemented in mlpack?
If not, can I start working on this?

**Comment by zoq** [comment]
Hello, if you like to work on this one please feel free, also please make sure to add a comprehensive test suite to the model implementation as well.

**Comment by saraansh1999** [comment]
Thanks. I'll start with it ASAP.

**Comment by saraansh1999** [comment]
Is there a grid sampler implemented in mlpack?

**Comment by kartikdutt18** [comment]
Hi @saraansh1999,
> 2\. Also once I start implementing it, will I have to make the whole of mlpack again and again to test any code?

Not really. You would have to run make command whenever make changes in codebase and it will only make the portion that you changed. When you are testing you simply make the tests. To run a particular test use, `./mlpack_test -t TestName`

> What will be the best place for implementing this? It is used as a module that can be just inserted into the beginning of the network, and is trained simply along the rest of the network without any special loss function or anything else.

For this a layer is independent of rest of the codebase however there are some constraints such as variables and the layers itself should be serializable. They must follow certain templated code as well. I think you read more about it in the documentation. I might be missing something so a member will give a better reply.

For the grid sampler, I don't think it is implemented but I am not 100% sure. I would probably need to read a bit more about grid samplers to comment on that. I hope that's okay.
Hope this helps.

**Comment by saraansh1999** [comment]
i have used the STN module multiple times in pytorch, but there we have the F.affine_grid and F.grid_sample functions available, which make the job very easy.

So my plan is somewhat as follows:
1. Implement the affine_grid_generator as a layer.
Reference: [https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/AffineGridGenerator.cpp](url)
2. Implement the grid_sampler as a layer.
Reference: [https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/GridSampler.cpp](url)
3. Implement the STN as a layer which takes as input a localization network and an input image. 
  3.1. Implement a forward function that calls upon the forward of the localization net, followed by the forwardo of the affine_grid_generator, followed by the forward of the grid_sample.
  3.2. Implement a backward function that calls upon the backward functions in the reverse order.
  3.3. Add a few checks like the output size of the localization net.
  3.4. Make tests.

A few questions / requests(again, sorry :) ):
1. It would be really helpful if someone could confirm if any of the above mentioned things are already present in mlpack. (I couldn't find them)
2. It would be helpful if someone could give a green light for the references listed above.
